Improve tests

LeonidBeynenson opened the pull request:
Improve tests a bit:
* Make NNCF tests work again
* Make each test write into its own output folder to decrease possibility of side effects between tests

LeonidBeynenson commented:
run tests

Ilya-Krylov commented on pytorch_toolkit/object_detection/tests/common/test_case.py:
```diff
@@ -65,235 +66,71 @@ def test_alt_ssd_export_on_cpu(self):
-            template_data.merge_from_dict(template_update_dict)
-            template_data.dump(template_file)
+            ap = [metric['value'] for metric in content['metrics'] if metric['key'] == 'bbox'][0]
+            ap = ap/100
```
need more spaces

LeonidBeynenson replied:
agree

LeonidBeynenson replied:
Fixed

Ilya-Krylov commented on pytorch_toolkit/ote/ote/tests/test_case.py:
```diff
@@ -165,18 +180,23 @@ def setUpClass(cls):
             download_snapshot_if_not_yet(cls.template_file, cls.template_folder)
 
+        def setUp(self):
+            self.output_folder = os.path.join(self.template_folder, f'output_{self.id()}')
```
please make appropriate changes in training_extensions/pytorch_toolkit/object_detection/tests/common/test_case.py on lines: 53, 59

LeonidBeynenson replied:
Oh! I see! thank you very much!

LeonidBeynenson replied:
Fixed

LeonidBeynenson commented on pytorch_toolkit/ote/ote/tests/test_case.py:
```diff
@@ -60,6 +63,10 @@ def skip_if_cpu_is_not_supported(template_file):
 
 def create_test_case(domain_name, problem_name, model_name, ann_file, img_root, metric_keys, expected_outputs_dir):
     class TestCaseOteApi(unittest.TestCase):
+        domain = domain_name
```
I will use it in my script for filtering tests by domain/topic/etc

Ilya-Krylov commented:
run tests

LeonidBeynenson commented:
2 @Ilya-Krylov , @druzhkov-paul , @evgeny-izutov :

After some considerations with other teams I decided to make our NNCF part of template files to be compliant with the external specs.
The changes are in my commit aad0726.

So, now the NNCF part looks like as follows:
```yaml
optimisations:
  nncf_quantization:
    config: "compression_config.json"
    default: false
  nncf_sparsity:
    config: "compression_config.json"
    default: false
```
Also I added the part that allows to switch on nncf compression algorithms (nncf_quantization/nncf_sparsity/etc) from the command line arguments using the cmd line parameters like `--nncf-quantization`, `--nncf-sparsity`, etc

The following changes are important
* I added to `api.py` the code that parses command line parameters such as `--nncf-quantization`
* Note that all the parameters has `action=store_true`, and the default value for them are read from the template.yaml, but if the value is False, the default value is `argparse.SUPPRESS`. 
  This is an important trick: if the default value is false, the argument does NOT appears in the parsed arguments at all!
* The command line arguments are passed to `compression_arg_transformer`, which makes all the work on generating NNCF part of mmdetection/mmaction/etc config file.
  Now it also 
  * removes unrequired arguments `nncf_quantization`, `nncf_sparsity` etc before passing them to trainer/evaluator/etc
  * returns if any of the compression flags were set -- at the moment we do not know about that until all command line arguments are converted
    (The latter allows to turn on/off compression after finetuning in the tool `train.py`)

LeonidBeynenson commented:
run tests

Ilya-Krylov commented on pytorch_toolkit/ote/ote/api.py:
```diff
@@ -2,6 +2,29 @@
+def _add_nncf_arguments_to_parser(parser, config):
+    def _get_def_val(nncf_section):
+        val = config.get('optimisations', {}).get(nncf_section, {}).get('default')
+        if not val:
```
I guess `if val is None:` must be here

Ilya-Krylov commented on pytorch_toolkit/ote/ote/api.py:
```diff
@@ -67,6 +95,7 @@ def test_args_parser(template_path):
                             help='Location where output images (with displayed result of model work) will be stored.')
         parser.add_argument('--config', default=config['config'], help=argparse.SUPPRESS)
 
+        _add_nncf_arguments_to_parser(parser, config)
```
I don't like adding this args to test and export script. From my perspective - keeping compression info in a snapshot is better.

Ilya-Krylov commented:
@LeonidBeynenson  please split this PR one two parts at least: test improvements and nncf enablement

LeonidBeynenson commented:
Split this PR into two PRs

This PR is the first one, it is small, just contains tests improvements.

The second PR is https://github.com/openvinotoolkit/training_extensions/pull/398
It contains all my changes to make NNCF part of template be consistent with external specs and to allow turning on the NNCF compression methods by command lines parameters such as `--nncf-quantization`, `--nncf-sparsity` , etc

LeonidBeynenson commented:
Add archive with log of all tests passing on my desktop 
[CI_log_006_2020-12-17_17-40-39.zip](https://github.com/openvinotoolkit/training_extensions/files/5715071/CI_log_006_2020-12-17_17-40-39.zip)

Ilya-Krylov commented on pytorch_toolkit/ote/ote/modules/compression/nncf_config_generator.py:
```diff
@@ -64,7 +66,11 @@ def __call__(self, template_path):
 
-        nncf_config_part = self._merge_nncf_compression_parts(compression_config_path, compression_parts_to_choose)
+        try:
+            nncf_config_part = self._merge_nncf_compression_parts(compression_config_path, compression_parts_to_choose)
```
why do you make this change?

LeonidBeynenson replied:
To avoid crash of the script.
Now if the template has inconsistent format, the script prints the error, compression is not working, but crash is not happened.

I can remove this if you thing that it is not required.

Ilya-Krylov replied:
please remove, in that case we will react faster if something goes wrong

LeonidBeynenson replied:
Fixed
